File: 11.盛最多水的容器.go
/*
 * @lc app=leetcode.cn id=11 lang=golang
 *
 * [11] 盛最多水的容器
 */

// @lc code=start
func maxArea(height []int) (cap int) {
	length := len(height)
	i, j := 0, length-1
	for i < j {
		area := min(height[i], height[j]) * (j - i)
		cap = max(cap, area)

		if height[i] < height[j] {
			i++
		} else {
			j--
		}
	}
	return
}
func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// @lc code=end

File: 11.盛最多水的容器.rs
/*
 * @lc app=leetcode.cn id=11 lang=rust
 *
 * [11] 盛最多水的容器
 */

// @lc code=start
impl Solution {
    pub fn max_area(height: Vec<i32>) -> i32 {
        let length = height.len();
        let (mut i, mut j) = (0, length - 1);
        let mut cap = 0;
        while i < j {
            let area = (height[i].min(height[j])) * (j - i) as i32;
            cap = cap.max(area);

            if height[i] < height[j] {
                i += 1;
            } else {
                j -= 1;
            }
        }
        cap
    }
}
// @lc code=end
